File: code/apps/web/src/lib/reception/mapping.ts
/**
 * Protobuf messages to the models the screens use.
 *
 * Everything above this file works in plain objects with string unions and
 * `Date`; everything below it is generated wire types with numeric enums and
 * protobuf `Timestamp`. Keeping the boundary in one tested module is what
 * stops an enum value being read as a truthy number somewhere in a template —
 * the kind of mistake that renders "Cancelled" as "Scheduled" and is invisible
 * in review.
 *
 * The translations are total. Every enum maps every case including
 * `UNSPECIFIED`, because a `default:` that silently picks a sensible-looking
 * value is how a new server enum arrives in the browser as something it is not.
 */
import { timestampDate } from '@bufbuild/protobuf/wkt';
import type { Timestamp } from '@bufbuild/protobuf/wkt';
import {
	DatePrecision as WireDatePrecision,
	IdentifierType as WireIdentifierType,
	MatchOutcome as WireMatchOutcome,
	PatientStatus as WirePatientStatus,
	Sex as WireSex,
	type Patient as WirePatient,
	type PatientMatch as WirePatientMatch
} from '$gen/healthcare/empi/v1/patient_pb.js';
import {
	AppointmentStatus as WireAppointmentStatus,
	ArrivalMode as WireArrivalMode,
	Priority as WirePriority,
	type Appointment as WireAppointment,
	type QueuePosition as WireQueuePosition
} from '$gen/healthcare/scheduling/v1/appointment_pb.js';
import {
	displayName,
	type BannerIdentifier,
	type DatePrecision,
	type PatientLike,
	type PatientStatus,
	type Sex
} from '../patient/banner.js';
import type { ArrivalMode, BoardAppointment, BoardPosition, Priority, QueueStatus } from './board.js';
import { presentMatch, type MatchOutcome, type PresentedMatch } from './search.js';

/** A protobuf Timestamp as a Date, or null when absent. */
export function toDate(timestamp: Timestamp | undefined): Date | null {
	return timestamp ? timestampDate(timestamp) : null;
}

const sexes: Record<WireSex, Sex> = {
	[WireSex.UNSPECIFIED]: 'unspecified',
	[WireSex.UNKNOWN]: 'unknown',
	[WireSex.FEMALE]: 'female',
	[WireSex.MALE]: 'male',
	[WireSex.OTHER]: 'other'
};

const precisions: Record<WireDatePrecision, DatePrecision> = {
	[WireDatePrecision.UNSPECIFIED]: 'unknown',
	[WireDatePrecision.DAY]: 'day',
	[WireDatePrecision.MONTH]: 'month',
	[WireDatePrecision.YEAR]: 'year',
	[WireDatePrecision.ESTIMATED]: 'estimated'
};

const statuses: Record<WirePatientStatus, PatientStatus> = {
	[WirePatientStatus.UNSPECIFIED]: 'unspecified',
	[WirePatientStatus.CANDIDATE]: 'candidate',
	[WirePatientStatus.ACTIVE]: 'active',
	[WirePatientStatus.MERGED]: 'merged',
	[WirePatientStatus.INACTIVE]: 'inactive'
};

const identifierLabels: Record<WireIdentifierType, string> = {
	[WireIdentifierType.UNSPECIFIED]: 'Identifier',
	[WireIdentifierType.MRN]: 'MRN',
	[WireIdentifierType.NATIONAL_HEALTH]: 'National health ID',
	[WireIdentifierType.GOVERNMENT]: 'Government ID',
	[WireIdentifierType.INSURANCE]: 'Insurance',
	[WireIdentifierType.EXTERNAL]: 'External ID'
};

const outcomes: Record<WireMatchOutcome, MatchOutcome> = {
	[WireMatchOutcome.UNSPECIFIED]: 'unspecified',
	[WireMatchOutcome.DISTINCT]: 'distinct',
	[WireMatchOutcome.REVIEW]: 'review',
	[WireMatchOutcome.PROBABLE]: 'probable',
	[WireMatchOutcome.CONFLICT]: 'conflict'
};

const queueStatuses: Record<WireAppointmentStatus, QueueStatus> = {
	// Unspecified is mapped to 'scheduled' rather than left to a default. It
	// only occurs for a server older than this client, and "expected" is the
	// state that offers a check-in — the wrong guess is recoverable, whereas
	// guessing "in consultation" hides a patient standing at the desk.
	[WireAppointmentStatus.UNSPECIFIED]: 'scheduled',
	[WireAppointmentStatus.SCHEDULED]: 'scheduled',
	[WireAppointmentStatus.ARRIVED]: 'arrived',
	[WireAppointmentStatus.TRIAGED]: 'triaged',
	[WireAppointmentStatus.WAITING_CLINICIAN]: 'waiting_clinician',
	[WireAppointmentStatus.IN_CONSULTATION]: 'in_consultation',
	[WireAppointmentStatus.POST_CONSULTATION]: 'post_consultation',
	[WireAppointmentStatus.COMPLETED]: 'completed',
	[WireAppointmentStatus.NO_SHOW]: 'no_show',
	[WireAppointmentStatus.CANCELLED]: 'cancelled'
};

const priorities: Record<WirePriority, Priority> = {
	// Unspecified becomes standard, which is what the server means by it, and
	// never 'immediate': a board that promotes every unset priority to the top
	// makes the real ones invisible.
	[WirePriority.UNSPECIFIED]: 'standard',
	[WirePriority.IMMEDIATE]: 'immediate',
	[WirePriority.VERY_URGENT]: 'very_urgent',
	[WirePriority.URGENT]: 'urgent',
	[WirePriority.STANDARD]: 'standard',
	[WirePriority.NON_URGENT]: 'non_urgent'
};

const arrivalModes: Record<WireArrivalMode, ArrivalMode> = {
	[WireArrivalMode.UNSPECIFIED]: 'scheduled',
	[WireArrivalMode.WALK_IN]: 'walk_in',
	[WireArrivalMode.SCHEDULED]: 'scheduled',
	[WireArrivalMode.AMBULANCE]: 'ambulance',
	[WireArrivalMode.REFERRAL]: 'referral',
	[WireArrivalMode.TELEHEALTH]: 'telehealth'
};

/** Adapts a wire Patient for the banner. */
export function toPatientLike(patient: WirePatient, masked = false): PatientLike {
	const demographics = patient.demographics;
	const birth = demographics?.birthDate;
	const birthDate = birth ? toDate(birth.date) : null;

	return {
		patientId: patient.patientId,
		status: statuses[patient.status] ?? 'unspecified',
		name: demographics?.name
			? { family: demographics.name.family, given: demographics.name.given }
			: null,
		birthDate: birthDate
			? { date: birthDate, precision: precisions[birth!.precision] ?? 'unknown' }
			: null,
		sex: sexes[demographics?.sex ?? WireSex.UNSPECIFIED] ?? 'unspecified',
		identifiers: patient.identifiers.map(
			(identifier): BannerIdentifier => ({
				label: identifierLabels[identifier.type] ?? 'Identifier',
				value: identifier.value,
				primary: identifier.primary
			})
		),
		deceased: patient.deceased
			? { date: patient.deceased.date ? toDate(patient.deceased.date.date) : null }
			: null,
		mergedIntoPatientId: patient.mergedIntoPatientId,
		designation: patient.designation
			? {
					label: patient.designation.label,
					apparentSex: sexes[patient.designation.apparentSex] ?? 'unspecified',
					apparentAge: patient.designation.apparentAge
				}
			: null,
		masked
	};
}

/** Adapts a wire PatientMatch for the search results. */
export function toPresentedMatch(match: WirePatientMatch): PresentedMatch {
	const patient = match.patient;
	const name = patient?.demographics?.name;
	const former = match.matchedFormerName?.name;

	return presentMatch({
		patientId: patient?.patientId ?? '',
		displayName: displayName(name ? { family: name.family, given: name.given } : null),
		confidence: match.confidence,
		outcome: outcomes[match.outcome] ?? 'unspecified',
		masked: match.masked,
		matchedFormerName: displayName(
			former ? { family: former.family, given: former.given } : null
		)
	});
}

/** Adapts a wire Appointment for the board. */
export function toBoardAppointment(appointment: WireAppointment): BoardAppointment {
	return {
		appointmentId: appointment.appointmentId,
		patientId: appointment.patientId,
		token: appointment.token,
		status: queueStatuses[appointment.status] ?? 'scheduled',
		priority: priorities[appointment.priority] ?? 'standard',
		priorityReason: appointment.priorityReason,
		arrivalMode: arrivalModes[appointment.arrivalMode] ?? 'scheduled',
		// A missing start time would sort to 1970 and put the row at the top of
		// the board. The epoch is at least obviously wrong; a silent `now` is
		// not.
		startsAt: toDate(appointment.startsAt) ?? new Date(0),
		checkedInAt: toDate(appointment.checkedInAt),
		version: appointment.version
	};
}

/** Adapts a wire QueuePosition for the board. */
export function toBoardPosition(position: WireQueuePosition): BoardPosition | null {
	if (!position.appointment) {
		// A position with no appointment cannot be rendered or acted on. Dropped
		// rather than rendered as a blank row, which reads as a patient whose
		// details failed to load.
		return null;
	}
	return {
		appointment: toBoardAppointment(position.appointment),
		position: position.position,
		estimatedWaitSeconds: Number(position.estimatedWaitSeconds)
	};
}

File: code/apps/mobile/lib/src/reception/mapping.dart
/// Wire types to the reception screens' models (UX-W1-01).
///
/// # How an enum this build has never heard of arrives
///
/// The web shell gets exhaustiveness for free: its maps are
/// `Record<WireEnum, T>` and `tsc` refuses a file that misses a member. That
/// caught a real mistake during UX-W1-03. Dart cannot do the same, because a
/// generated protobuf enum is a class of constants rather than a closed set.
///
/// The obvious runtime substitute — a `default:` arm that throws — is worse
/// than useless here, because it can never run. Measured rather than assumed:
/// `protobuf.dart` decodes an unrecognised enum tag as the **zero member** and
/// files the real tag under `unknownFields`. So a future
/// `MATCH_OUTCOME_MERGED` does not arrive as something unmapped; it arrives as
/// `MATCH_OUTCOME_UNSPECIFIED`, which this build reads as "match strength not
/// assessed" — a verdict that does not block registration. A newer server
/// would silently unlock the create button on a patient already in the index.
///
/// So the check is on `unknownFields`, which is where the evidence actually
/// is. A field whose tag appears there was sent as a value this build cannot
/// name, and the mapping returns an explicit `unrecognised` member rather than
/// the zero one. Each of those members is defined to fail safe: an
/// unrecognised match outcome blocks registration, an unrecognised appointment
/// status offers no check-in.
///
/// Refusing the whole response instead would be defensible, but it turns one
/// unfamiliar row into an unusable screen, and a receptionist with a queue in
/// front of them will find another way. One marked row is the better failure.
library;

import 'package:protobuf/protobuf.dart' as pb;

import '../gen/healthcare/empi/v1/patient.pb.dart' as empi;
import '../gen/healthcare/scheduling/v1/appointment.pb.dart' as sched;
import 'board.dart';
import 'search.dart';

/// Field tags carrying the enums this module reads.
///
/// Named constants because the number is the whole check: reading the wrong
/// tag would silently stop detecting anything, and a test pins each one.
const int matchOutcomeField = 3; // empi.v1.PatientMatch.outcome
const int appointmentStatusField = 10; // scheduling.v1.Appointment.status
const int arrivalModeField = 21; // scheduling.v1.Appointment.arrival_mode
const int priorityField = 23; // scheduling.v1.Appointment.priority

/// True when [message] carried a value for [tag] that this build could not
/// interpret.
bool sentUnknownValueFor(pb.GeneratedMessage message, int tag) =>
    message.unknownFields.hasField(tag);

/// Match outcome, wire to model.
///
/// Takes the enclosing message rather than the enum, because the enum alone
/// cannot answer the question — an unknown value and a genuinely unspecified
/// one are the same constant by the time they reach a getter.
MatchOutcome matchOutcomeOf(empi.PatientMatch wire) {
  if (sentUnknownValueFor(wire, matchOutcomeField)) {
    return MatchOutcome.unrecognised;
  }
  return switch (wire.outcome) {
    empi.MatchOutcome.MATCH_OUTCOME_DISTINCT => MatchOutcome.distinct,
    empi.MatchOutcome.MATCH_OUTCOME_REVIEW => MatchOutcome.review,
    empi.MatchOutcome.MATCH_OUTCOME_PROBABLE => MatchOutcome.probable,
    empi.MatchOutcome.MATCH_OUTCOME_CONFLICT => MatchOutcome.conflict,
    empi.MatchOutcome.MATCH_OUTCOME_UNSPECIFIED => MatchOutcome.unspecified,
    // Unreachable by the decoding above, and kept rather than left to an
    // analyser warning: if protobuf's behaviour ever changes, failing safe is
    // still the right answer.
    _ => MatchOutcome.unrecognised,
  };
}

/// Appointment status, wire to model.
QueueStatus queueStatusOf(sched.Appointment wire) {
  if (sentUnknownValueFor(wire, appointmentStatusField)) {
    return QueueStatus.unrecognised;
  }
  return switch (wire.status) {
    sched.AppointmentStatus.APPOINTMENT_STATUS_SCHEDULED => QueueStatus.scheduled,
    sched.AppointmentStatus.APPOINTMENT_STATUS_ARRIVED => QueueStatus.arrived,
    sched.AppointmentStatus.APPOINTMENT_STATUS_TRIAGED => QueueStatus.triaged,
    sched.AppointmentStatus.APPOINTMENT_STATUS_WAITING_CLINICIAN =>
      QueueStatus.waitingClinician,
    sched.AppointmentStatus.APPOINTMENT_STATUS_IN_CONSULTATION =>
      QueueStatus.inConsultation,
    sched.AppointmentStatus.APPOINTMENT_STATUS_POST_CONSULTATION =>
      QueueStatus.postConsultation,
    sched.AppointmentStatus.APPOINTMENT_STATUS_COMPLETED => QueueStatus.completed,
    sched.AppointmentStatus.APPOINTMENT_STATUS_NO_SHOW => QueueStatus.noShow,
    sched.AppointmentStatus.APPOINTMENT_STATUS_CANCELLED => QueueStatus.cancelled,
    sched.AppointmentStatus.APPOINTMENT_STATUS_UNSPECIFIED =>
      QueueStatus.unrecognised,
    _ => QueueStatus.unrecognised,
  };
}

/// Priority, wire to model.
///
/// An unspecified priority is standard, which the proto says in as many words
/// — a documented meaning rather than a guess, so it maps. A tag this build
/// cannot name is a different thing and gets its own member.
QueuePriority queuePriorityOf(sched.Appointment wire) {
  if (sentUnknownValueFor(wire, priorityField)) {
    return QueuePriority.unrecognised;
  }
  return switch (wire.priority) {
    sched.Priority.PRIORITY_IMMEDIATE => QueuePriority.immediate,
    sched.Priority.PRIORITY_VERY_URGENT => QueuePriority.veryUrgent,
    sched.Priority.PRIORITY_URGENT => QueuePriority.urgent,
    sched.Priority.PRIORITY_STANDARD => QueuePriority.standard,
    sched.Priority.PRIORITY_NON_URGENT => QueuePriority.nonUrgent,
    sched.Priority.PRIORITY_UNSPECIFIED => QueuePriority.standard,
    _ => QueuePriority.unrecognised,
  };
}

/// Arrival mode, wire to model.
ArrivalMode arrivalModeOf(sched.Appointment wire) {
  if (sentUnknownValueFor(wire, arrivalModeField)) {
    return ArrivalMode.unrecognised;
  }
  return switch (wire.arrivalMode) {
    sched.ArrivalMode.ARRIVAL_MODE_WALK_IN => ArrivalMode.walkIn,
    sched.ArrivalMode.ARRIVAL_MODE_SCHEDULED => ArrivalMode.scheduled,
    sched.ArrivalMode.ARRIVAL_MODE_AMBULANCE => ArrivalMode.ambulance,
    sched.ArrivalMode.ARRIVAL_MODE_REFERRAL => ArrivalMode.referral,
    sched.ArrivalMode.ARRIVAL_MODE_TELEHEALTH => ArrivalMode.telehealth,
    sched.ArrivalMode.ARRIVAL_MODE_UNSPECIFIED => ArrivalMode.scheduled,
    _ => ArrivalMode.unrecognised,
  };
}

/// One appointment, wire to model.
BoardAppointment appointmentOf(sched.Appointment wire) => BoardAppointment(
      appointmentId: wire.appointmentId,
      patientId: wire.patientId,
      token: wire.token,
      status: queueStatusOf(wire),
      priority: queuePriorityOf(wire),
      priorityReason: wire.priorityReason,
      arrivalMode: arrivalModeOf(wire),
      startsAt: wire.startsAt.toDateTime().toUtc(),
      // hasCheckedInAt rather than a zero-timestamp check: the epoch is a real
      // instant, and "not checked in" has to stay distinguishable from a clock
      // that was wrong.
      checkedInAt:
          wire.hasCheckedInAt() ? wire.checkedInAt.toDateTime().toUtc() : null,
      version: wire.version.toInt(),
    );

/// One queue position, wire to model.
BoardPosition positionOf(sched.QueuePosition wire) => BoardPosition(
      appointment: appointmentOf(wire.appointment),
      position: wire.position,
      estimatedWaitSeconds: wire.estimatedWaitSeconds.toInt(),
    );

/// One search candidate, wire to model.
PresentedMatch matchOf(empi.PatientMatch wire) {
  // No fallback here, unlike the display name: an absent former name is
  // absent, and "also known as Name not available" is worse than silence.
  final formerName =
      wire.hasMatchedFormerName() ? _join(wire.matchedFormerName.name) : '';

  return presentMatch(
    patientId: wire.patient.patientId,
    // An empty name is shown as such rather than as a blank row. A row with
    // nothing in the name column reads as a rendering fault, and a
    // receptionist scanning for "is this them" skips it.
    displayName: _orUnavailable(_join(wire.patient.demographics.name)),
    confidence: wire.confidence,
    outcome: matchOutcomeOf(wire),
    masked: wire.masked,
    matchedFormerName: formerName,
  );
}

/// Joins a name's parts, which may legitimately produce nothing.
String _join(empi.HumanName name) => [name.given.join(' ').trim(), name.family]
    .where((p) => p.isNotEmpty)
    .join(' ');

/// The fallback for a name a row has to show something for.
String _orUnavailable(String name) =>
    name.isEmpty ? 'Name not available' : name;
